pkg/controllers/route_controller.go

```diff
@@ -456,3 +430,196 @@ func (r *routeReconciler) validateBackendRefsIpFamilies(ctx context.Context, rou
 
 	return nil
 }
+
+var (
+	ErrValidation          = errors.New("validation")
+	ErrParentRefsNotFound  = errors.New("parentRefs are not found")
+	ErrRouteGKNotSupported = errors.New("route GroupKind is not supported")
+)
+
+// Validation for route spec. Will validate and update route status. Returns error if not valid.
+// Validation rules are suppose to be compliant to Gateway API Spec.
+//
+//	https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1.RouteConditionType
+//
+// There are 3 condition types: Accepted, PartiallyInvalid, ResolvedRefs.
+// We dont support PartiallyInvalid for now and reject entire route if there is at least one invalid field.
+// Accepted type is related to parentRefs, and ResolvedRefs to backendRefs. These 2 are validated independently.
+func (r *routeReconciler) validateRoute(ctx context.Context, route core.Route) error {
+	ps, err := r.validateRouteParentRefs(ctx, route)
+	if err != nil {
+		return err
+	}
+
+	cnd, err := r.validateBackedRefs(ctx, route)
+	if err != nil {
+		return err
+	}
+
+	// we need to update each parentRef with backendRef status
+	psbr := make([]gwv1.RouteParentStatus, len(ps))
+	for i, rps := range ps {
+		meta.SetStatusCondition(&rps.Conditions, cnd)
+		psbr[i] = rps
+	}
+
+	route.Status().SetParents(psbr)
+
+	err = r.client.Status().Update(ctx, route.K8sObject())
+	if err != nil {
+		return fmt.Errorf("validate route: %w", err)
+	}
+
+	if r.hasNotAcceptedCondition(route) {
+		return fmt.Errorf("%w: route has validation errors, see status", ErrValidation)
+	}
+
+	return nil
+}
+
+// checks if route has at least single condition with status = false
+func (r *routeReconciler) hasNotAcceptedCondition(route core.Route) bool {
+	rps := route.Status().Parents()
+	for _, ps := range rps {
+		for _, cnd := range ps.Conditions {
+			if cnd.Status != metav1.ConditionTrue {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// find Gateway by Route and parentRef, returns nil if not found
+func (r *routeReconciler) findRouteParentGw(ctx context.Context, route core.Route, parentRef gwv1beta1.ParentReference) (*gwv1beta1.Gateway, error) {
+	ns := route.Namespace()
+	if parentRef.Namespace != nil && *parentRef.Namespace != "" {
+		ns = string(*parentRef.Namespace)
+	}
+	gwName := types.NamespacedName{
+		Namespace: ns,
+		Name:      string(parentRef.Name),
+	}
+	gw := &gwv1beta1.Gateway{}
+	err := r.client.Get(ctx, gwName, gw)
+	if err != nil {
+		return nil, client.IgnoreNotFound(err)
+	}
+	return gw, nil
+}
+
+// Validation rules for route parentRefs
+//
+// Will ignore status update when:
+// - parentRef does not exists, includes when parentRef Kind is not Gateway
+//
+// If parent GW exists will check:
+// - NoMatchingParent: parentRef sectionName and port matches Listener name and port
+// - TODO: NoMatchingListenerHostname: listener hostname matches one of route hostnames
+// - TODO: NotAllowedByListeners: listener allowedRoutes contains route GroupKind
```

Comment:
Do we have action items associated with these TODOs?

Reply:
yes